pkg/s3select/select.go

```diff
@@ -306,22 +316,30 @@ func (s3Select *S3Select) Open(getReader func(offset, length int64) (io.ReadClos
 	panic(fmt.Errorf("unknown input format '%v'", s3Select.Input.format))
 }
 
-func (s3Select *S3Select) marshal(record sql.Record) ([]byte, error) {
+func (s3Select *S3Select) marshal(b *bytes.Buffer, bufW *bufio.Writer, record sql.Record) error {
 	switch s3Select.Output.format {
 	case csvFormat:
-		data, err := record.MarshalCSV([]rune(s3Select.Output.CSVArgs.FieldDelimiter)[0])
+		// Use bufio Writer to prevent csv.Writer from allocating new buffer.
+		bufW.Reset(b)
+		err := record.WriteCSV(bufW, []rune(s3Select.Output.CSVArgs.FieldDelimiter)[0])
 		if err != nil {
-			return nil, err
+			return err
 		}
 
-		return append(data, []byte(s3Select.Output.CSVArgs.RecordDelimiter)...), nil
+		b.Truncate(b.Len() - 1)
+		b.WriteString(s3Select.Output.CSVArgs.RecordDelimiter)
+
+		return nil
 	case jsonFormat:
-		data, err := record.MarshalJSON()
+		err := record.WriteJSON(b)
 		if err != nil {
-			return nil, err
+			return err
 		}
 
-		return append(data, []byte(s3Select.Output.JSONArgs.RecordDelimiter)...), nil
+		b.Truncate(b.Len() - 1)
```

Comment:
You are skipping `}` closing of JSON.

Reply:
Here the WriteJson using `json.NewEncoder(w).Encode(r.KVS)` instead of the old way `json.Marshal(r.KVS)`
and '\n' is appended, will need to be removed.

https://github.com/golang/go/blob/master/src/encoding/json/stream.go#L213

Reply:
Got it